nerfstudio/model_components/renderers.py

```diff
@@ -124,65 +128,68 @@ def get_background_color(
+        background_color: Optional[Float[Tensor, "3"]] = None,
     ) -> Float[Tensor, "*bs 3"]:
-        """Blends the background color into the image.
+        """Blends the background color into the image if image is RGBA.
+        Otherwise no blending is performed (we assume opacity of 1).
 
         Args:
-            rgb: RGB per pixel.
+            image: RGB/RGBA per pixel.
             opacity: Alpha opacity per pixel.
             background_color: Background color.
 
         Returns:
             Blended RGB.
         """
-        if len(opacity.shape) < len(rgb.shape):
-            opacity = opacity.unsqueeze(-1)
+        if background_color is None:
+            background_color_name = self.background_color
+            if background_color_name == "last_sample" or background_color == "random":
+                background_color_name = "black"
+            background_color = self.get_background_color(
+                rgb=image.unsqueeze(-1), background_color=background_color_name
+            )
+        if image.size(-1) < 4:
+            return image
+        rgb, opacity = image[..., :3], image[..., 3:]
         blended_rgb = rgb * opacity + background_color.to(rgb.device) * (1 - opacity)
         return blended_rgb
 
-    def blend_background_for_metric_computation(
+    def blend_background_for_loss_computation(
         self,
+        pred_image: Tensor,
+        pred_accumulation: Tensor,
         gt_image: Tensor,
-        predicted_rgb: Tensor,
-        predicted_opacity: Tensor,
-        background_color: Float[Tensor, "3"] = colors.BLACK,
     ) -> Tuple[Tensor, Tensor]:
         """Blends a background color into the ground truth and predicted image for
-        metric computation if the current background color is "random".
+        loss computation.
 
         Args:
             gt_image: The ground truth image.
-            predicted_rgb: The predicted RGB values.
-            predicted_opacity: The predicted opacity/ accumulation.
-            background_color: The background color to use for blending. This should be
-                a single color to use for all pixels.
-
+            pred_image: The predicted RGB values (without background blending).
+            pred_accumulation: The predicted opacity/ accumulation.
         Returns:
             A tuple of the predicted and ground truth RGB values.
         """
-        if gt_image.shape[-1] == 4 and self.background_color == "random":
-            predicted_rgb = self.blend_background(
-                rgb=predicted_rgb, opacity=predicted_opacity, background_color=background_color
-            )
-            gt_rgb = self.blend_background(
-                rgb=gt_image[..., :3], opacity=gt_image[..., 3], background_color=background_color
-            )
-        else:
-            gt_rgb = gt_image[..., :3]
 
-        return predicted_rgb, gt_rgb
+        if self.background_color == "random":
+            background_color = torch.rand_like(pred_image)
```

Comment:
Why move this here rather than leaving it in `get_background_color`? The "last_sample" color also isn't constant

Reply:
I will leave it there. I was a bit afraid people would use get_background_color for their own stuff without being aware of the risks.

Reply:
I see what you mean. Maybe `get` is a bad name. What about renaming it to `compute`, making it a bit clearer that the output might be different everytime that it is called?

Reply:
Did you introduce this function or was it in nerfstudio before?

Reply:
I introduced it. It used to be part of `combine_rgb`

Reply:
Great, then we are free to define it as we see fit. I will do some refactoring.